Optional long-polling based segment announcement via HTTP instead of Zookeeper

himanshug opened the pull request:
This PR  introduces following configurations.

```
//to be used at coordinators and brokers to use an implementation of ServerInventoryView that
//syncs segments list over HTTP instead of zookeeper
druid.announcer.type=http

//to be used at historicals and realtime tasks to stop pushing segment updates to zookeeper
//false by default
druid.announcer.skipSegmentAnnouncementOnZk=true
```

I'm keeping those undocumented for now. We will use them on our big internal clusters first before asking users to use.

changes -
All nodes, that serve segments e.g. historicals and realtime indexing tasks, provide a "/druid-internal/v1/segments" HTTP endpoint that can incrementally provide the list of segments being served by that node, this uses Async IO so jetty threads are not held waiting. To power this endpoint, following method is added to DataSegmentAnnouncer interface.

```
  /**
   * Returns Future that lists the segment load/drop requests since given counter.
   */
  public ListenableFuture<SegmentChangeRequestsSnapshot> getSegmentChangesSince(SegmentChangeRequestHistory.Counter counter);
```

On coordinator/brokers side, HttpServerInventoryView class is introduced which is equivalent to BatchServerInventoryView but syncs the segment inventory using the mentioned HTTP endpoint. Server discovery
is still done via zookeeper. ServerInventoryView is made an interface and its old contents are moved to AbstractCuratorServerInventoryView .


TODO: (subsequent PRs)
- reserve some jetty resources for "/druid-internal/v1/segments" endpoint
- In HttpServerInventoryView, switch the server discovery also to HTTP and make it pure HTTP.

kaijianding commented:
If I understand correctly, you update segments by calling httpClient.go(url) for each server in the `servers` list infinitely inside a thread pool.
To my knowledge, the httpClient.go() use a connection pool which actually using ImmediateCreationResourceHolder. This resourceHolder holds an objectList whose size is configed by broker.numConnections. httpClient.go takes first connection from the head of objectList and give it back and add the connection to the end of objectList when httpClient.go() finished.
In this way all connections holed by this resourceHolder are active(or hot?), whether this is a kind of waste if the druid cluster is a not a busy cluster?

If the cluster is not a busy cluster, can a connection is in `disconnected` state(disconnected by historical/realtime node when over the `idleTimeOut`) when attempt to use this connection to get segment list? In this case a new connection is created(renew), will it cause performance issue if lots of connections need to be renew?

himanshug commented:
@kaijianding thanks for taking a look. you are right about the connection usage and yes coordinator/broker would be using one connection from HttpClient per historical/realtime all the time.
On a non-busy cluster or in the worst case coordinator/broker would have to create one connection per historical/realtime every 4 mins ( default config in HttpServerInventoryViewConfig ). But, They will be created sequentially and not in parallel.
However, it turns out that this problem can be very easily fixed by changing the get/giveBack behavior of ResourcePool.ImmediateCreationResourceHolder to LIFO instead of current FIFO. It can be done by replacing https://github.com/metamx/http-client/blob/master/src/main/java/com/metamx/http/client/pool/ResourcePool.java#L237 by `objectList.addFirst(object)`. I will do a PR to fix that.

btw, I have also made few updates in SegmentListerResource to make it async and not hold any jetty threads while in wait.

drcrallen commented:
Related https://github.com/druid-io/druid/pull/2368

himanshug commented:
removing the discovery of historical/realtime nodes from zookeeper is gonna relate with https://github.com/druid-io/druid/issues/2312 (as described in https://github.com/druid-io/druid/issues/2312 ) . I am looking into that and that would be done in a separate PR.

weijietong commented:
@himanshug  according to your idea, through looping asking segments from historical/realtime nodes background ,how to avoid holding stale segment infos on coordinator/brokers side if not looped in time while a query occurred ?

himanshug commented:
@weijietong brokers/coordinators would not have stale segment info because they are running inifinite (no wait) loop to get latest segment information from historical/realtime nodes.... historical/realtime nodes "hold" the request till there is new information to provide or timeout provided in the request is reached.

pjain1 commented:
@himanshug a general comment - it would be good to put some comments on important classes, variables and methods. For example, a high level design comment on `HttpServerInventoryView`, some comments on `SegmentChangeRequestHistory`, usage of `hash` in `Counter`, reason of passing `waitingFutures` in `SettableFuture` etc. This would make it easier to understand the code and help the person changing the code not make mistakes.

I have gone through the code at high level and it looks good me to so far.

himanshug commented:
@pjain1 added some docs for clarification

gianm commented:
@himanshug, I know I said on the dev sync today that I thought this should remain undocumented, but I changed my mind. I'm still ok with it only being reviewed by people from your organization, but I think it should be documented as experimental. That way other folks could help try it out too.

cheddar commented on server/src/main/java/io/druid/client/HttpServerInventoryView.java:
```diff
@@ -0,0 +1,719 @@
+      final @Global HttpClient httpClient,
+      final CuratorFramework curator,
+      final String announcementsPath,
+      final Predicate<Pair<DruidServerMetadata, DataSegment>> defaultFilter,
```
`Pair` is a horrible class to use in actual public APIs or any other API contract.  The reason is because it can quickly become difficult to know what the lhs and rhs are actually representing.  As a general code design rule, it's better to only use `Pair` as a convenience for internal things and if you are actually creating a contract to create a class that represents your pair of items with actual names.

cheddar commented on server/src/main/java/io/druid/client/HttpServerInventoryView.java:
```diff
@@ -0,0 +1,719 @@
+
+        Futures.addCallback(
+            future,
+            new FutureCallback<InputStream>()
```
I'm actually not sure which thread this callback will be called on, but I think that it might actually block in some not good locations (i.e. the Future will be done and returned when the first response packet comes back, but will block waiting on any subsequent chunks).  Rather than doing a `Future` implementation, I think it would be better to implement our own `ResponseHandler` here.  It should return `incomplete` responses essentially building up a BAOS or something similar until the final chunk, where it can convert that into the snapshot, do the work and finally return a `complete` response.

himanshug replied:
gets called in netty thread that completes the future. Implemented BytesAccumulatingResponseHandler on top of existing InputStreamResponseHandler though to collect all chunks.

cheddar commented on server/src/main/java/io/druid/client/ServerInventoryView.java:
```diff
@@ -19,329 +19,12 @@
 /**
  */
-public abstract class ServerInventoryView<InventoryType> implements ServerView, InventoryView
+public interface ServerInventoryView extends ServerView, InventoryView
```
Given that this is just 2 more methods, how useful is the interface?  Do we really need it?

himanshug replied:
moved the methods to InventoryView , however marker interface is still used to keep it configurable.

cheddar commented on server/src/main/java/io/druid/server/coordination/BatchDataSegmentAnnouncer.java:
```diff
@@ -95,18 +103,37 @@ public DataSegment apply(DataSegment input)
       }
     };
+
+    if (this.config.isSkipSegmentAnnouncementOnZk()) {
```
Instead of the various changes in thsi class, it might be nice to have a different `DataSegmentAnnouncer` that essentially just adds to the lookups and has `getSegmentChangesSince` on the concrete type rather than the interface.  We then create a `CompositeDataSegmentAnnouncer` which can take both of them and call both.  The endpoint would have only the concrete lookup type injected into it (marked @Nullable and intelligent about returning a `404` if the object is `null`).

himanshug replied:
history keeper needs access to list of all segments served by the node to support first segment list call (after that deltas are delivered). this list of segments is already maintained in-memory inside BatchDataSegmentAnnouncer and I did not want to clone that list in another place.
Removed `getSegmentChangesSince(..)` from `DataSegmentAnnouncer` interface and made `SegmentListerResource` directly use `BatchDataSegmentAnnouncer`

cheddar commented on server/src/main/java/io/druid/server/coordination/SegmentChangeRequestHistory.java:
```diff
@@ -0,0 +1,300 @@
+    this(MAX_SIZE);
+  }
+
+  @VisibleForTesting
```
I don't think this is only visible for testing.  Makes a ton of sense to me for this to be part of the public API.

himanshug replied:
made it public

cheddar commented on server/src/main/java/io/druid/server/coordination/SegmentChangeRequestHistory.java:
```diff
@@ -0,0 +1,300 @@
+   * Returns a Future that , on completion, returns list of segment updates and associated counter.
+   * If there are no update since given counter then Future completion waits till an updates is provided.
+   *
+   * If counter is very older than max number of changes maintained than IAE exception is thrown.
```
I'm not sure that an exception is correct.  It might be nicer to have the `SegmentChangeRequestsSnapshot` have some indication that it is empty or invalid because the counter was too old.

himanshug replied:
ok,  now using a `resetCounter` boolean variable in `SegmentChangeRequestsSnapshot` instead of throwing exception when counter is older than the kept max history.

cheddar commented on server/src/main/java/io/druid/server/coordination/SegmentChangeRequestHistory.java:
```diff
@@ -0,0 +1,300 @@
+  {
+    for (DataSegmentChangeRequest request : requests) {
+      if (changes.size() >= maxSize) {
+        changes.remove(0);
```
This might be pre-mature optimization, but I worry about how this data structure is being used.  One option would be to view the list as a revolving buffer and essentially maintain an index of the oldest "valid" item.  You can replace and increment that on `add` and start your search from there wrapping back around.

himanshug replied:
added a `CircularBuffer` impl instead of using list as a circular buffer.

cheddar commented on server/src/main/java/io/druid/server/coordination/SegmentChangeRequestHistory.java:
```diff
@@ -0,0 +1,300 @@
+    Counter lastCounter = getLastCounter();
+
+    if (counter.counter >= lastCounter.counter) {
+      throw new IAE("counter >= last counter");
```
Please include the actual values that came in here, just the message as it exists won't really help much with debugging if it actually gets thrown.

himanshug replied:
updated

cheddar commented on server/src/main/java/io/druid/server/coordination/SegmentChangeRequestHistory.java:
```diff
@@ -0,0 +1,300 @@
+    } else {
+      int changeStartIndex = (int) (counter.counter + changes.size() - lastCounter.counter);
+
+      Counter counterToMatch = counter.counter == 0 ? Counter.ZERO : changes.get(changeStartIndex - 1).counter;
```
This is an optimization, maybe not so important, but I *think* you could actually compute the index into the array with some maths and the counters.

himanshug replied:
not really and also as you said not so important.

cheddar commented on server/src/main/java/io/druid/server/coordination/ZkCoordinator.java:
```diff
@@ -349,13 +349,11 @@ each time when addSegment() is called, it has to wait for the lock in order to m
-          throw new SegmentLoadingException(e, "Failed to announce segment[%s]", segment.getIdentifier());
-        }
+      try {
+        announcer.announceSegment(segment);
```
Why are we removing the `isAnnounced` check?

himanshug replied:
updated BatchDataSegmentAnnouncer.announce(..) to always do the check so that it gets done in other places as well.

cheddar commented on server/src/main/java/io/druid/server/http/SegmentListerResource.java:
```diff
@@ -0,0 +1,218 @@
+      @Json ObjectMapper jsonMapper,
+      @Smile ObjectMapper smileMapper,
+      AuthConfig authConfig,
+      DataSegmentAnnouncer announcer
```
If you do what I wrote in a previous comment and separate the lookup management into a different implementation of `DataSegmentAnnouncer`, then have that concrete type injected here instead of the interface (if you mark it `@Nullable` Guice should be fine even if it's not bound because the http thing isn't being used)

himanshug replied:
this now directly uses BatchDataSegmentAnnouncer

himanshug commented on server/src/main/java/io/druid/client/HttpServerInventoryView.java:
```diff
@@ -0,0 +1,719 @@
+      final @Json ObjectMapper jsonMapper,
+      final @Smile ObjectMapper smileMapper,
+      final @Global HttpClient httpClient,
+      final CuratorFramework curator,
```
refactored server notifications in separate class .

himanshug commented on server/src/main/java/io/druid/client/HttpServerInventoryView.java:
```diff
@@ -0,0 +1,719 @@
+
+    log.info("Calling SegmentCallback.segmentViewInitialized() for all callbacks.");
+
+    runSegmentCallbacks(
```
yes , it is for individual segments and once at inventory initialization. this naming and behavior is retained from CuratorBasedServerInventoryView ( called ServerInventoryView earlier)

himanshug commented on server/src/main/java/io/druid/client/HttpServerInventoryView.java:
```diff
@@ -0,0 +1,719 @@
+    if (curr == null) {
+      try {
+        Future<?> future = newHolder.updateSegmentsListAsync();
+        future.get(config.getServerTimeout(), TimeUnit.MILLISECONDS);
```
not needed anymore , it was needed in an older version of code ... removed now

himanshug commented on server/src/main/java/io/druid/client/HttpServerInventoryView.java:
```diff
@@ -0,0 +1,719 @@
+  public boolean isSegmentLoadedByServer(String serverKey, DataSegment segment)
+  {
+    synchronized (servers) {
+      DruidServerHolder holder = servers.get(serverKey);
```
added check

himanshug commented on server/src/main/java/io/druid/client/HttpServerInventoryView.java:
```diff
@@ -0,0 +1,719 @@
+  {
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
```
added comment in code, this is to ensure consistent segment state stored in DruidServer and counter managed.

himanshug commented:
@cheddar I also introduced `DataSegmentServerAnnouncer` and removed server announcements from `DataSegmentAnnouncer` implementation. Now only node who really serve segments will announce themselves in inventory and not random peon task processes.

himanshug commented:
@gianm I still think we should keep it undocumented for now and not confuse regular users with a feature which is just to optimize things and haven't been used at scale yet. We will try and document it in next release probably.

cheddar commented:
👍
